File: core/jvm/main/scala/com/webank/eggroll/core/command/CommandModel.scala
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.webank.eggroll.core.command

import java.net.{URI, URLDecoder}
import java.nio.charset.StandardCharsets

import com.google.protobuf.{ByteString, Message => PbMessage}
import com.webank.eggroll.core.constant.{ObjectConstants, StringConstants}
import com.webank.eggroll.core.rpc.RpcMessage
import com.webank.eggroll.core.serdes.{PbMessageDeserializer, PbMessageSerializer}
import org.apache.commons.lang3.StringUtils

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.reflect.ClassTag

case class ErCommandRequest(seq: Long = System.currentTimeMillis(), uri: String, args: Array[Array[Byte]] = null, kwargs: mutable.Map[String, Array[Byte]] = null) extends RpcMessage

case class ErCommandResponse(seq: Long, request: ErCommandRequest = null, data: Array[Byte] = null) extends RpcMessage

class CommandURI(uriString: String) {
  val uri = new URI(uriString)
  val queryString = uri.getQuery
  private val queryPairs = mutable.Map[String, String]()

  def this(src: ErCommandRequest) {
    this(src.uri)
  }

  def this(src: ErCommandResponse) {
    this(src.request.uri)
  }

  if (StringUtils.isBlank(queryString)) {
    queryPairs.put(StringConstants.ROUTE, uriString)
  } else {
    for (pair <- queryString.split(StringConstants.AND)) {
      val idx = pair.indexOf(StringConstants.EQUAL)
      val key = if (idx > 0) URLDecoder.decode(pair.substring(0, idx), StandardCharsets.UTF_8.name()) else pair
      val value = if (idx > 0 && pair.length > idx + 1) URLDecoder.decode(pair.substring(idx + 1), StandardCharsets.UTF_8.name()) else StringConstants.EMPTY
      queryPairs.put(key, value)
    }
  }

  def getQueryValue(key: String): String = {
    queryPairs(key)
  }

  def getRoute(): String = {
    queryPairs(StringConstants.ROUTE)
  }
}

object CommandPbSerdes {

  implicit class ErCommandRequestToPbMessage(src: ErCommandRequest) extends PbMessageSerializer {
    override def toProto[T >: PbMessage](): Command.CommandRequest = {
      val builder = Command.CommandRequest.newBuilder()
        .setSeq(src.seq)
        .setUri(src.uri)

      if (src.args != null) {
        src.args.map(arg => builder.addArgs(ByteString.copyFrom(arg)))
      }
      if (src.kwargs != null) {
        src.kwargs.map(kwarg => builder.putKwargs(kwarg._1, ByteString.copyFrom(kwarg._2)))
      }
      builder.build()
    }
  }

  implicit class ErCommandResponseToPbMessage(src: ErCommandResponse) extends PbMessageSerializer {
    override def toProto[T >: PbMessage](): Command.CommandResponse = {
      val builder = Command.CommandResponse.newBuilder()
        .setSeq(src.seq)
        .setRequest(if (src.request == null) Command.CommandRequest.getDefaultInstance else src.request.toProto())
        .setData(ByteString.copyFrom(src.data))

      builder.build()
    }
  }

  implicit class ErCommandRequestFromPbMessage(src: Command.CommandRequest) extends PbMessageDeserializer {
    override def fromProto[T >: RpcMessage](): ErCommandRequest = {
      var args = ObjectConstants.EMPTY_ARRAY_OF_BYTE_ARRAY
      val argsCount = src.getArgsCount
      if (argsCount > 0) {
        val argsBuffer = new ArrayBuffer[Array[Byte]]()

        src.getArgsList.forEach(bs => argsBuffer.append(bs.toByteArray))

        args = argsBuffer.toArray
      }

      var kwargs = ObjectConstants.EMPTY_MUTABLE_MAP_OF_STRING_TO_BYTE_ARRAY
      if (src.getKwargsCount > 0) {
        kwargs = mutable.Map[String, Array[Byte]]()

        src.getKwargsMap.entrySet().forEach(entry => kwargs.put(entry.getKey, entry.getValue.toByteArray))
      }

      ErCommandRequest(seq = src.getSeq,
        uri = src.getUri,
        args = args,
        kwargs = kwargs)
    }

    override def fromBytes[T: ClassTag](bytes: Array[Byte]): T = ???
  }

  implicit class ErCommandResponseFromPbMessage(src: Command.CommandResponse) extends PbMessageDeserializer {
    override def fromProto[T >: RpcMessage](): ErCommandResponse = ErCommandResponse(
      seq = src.getSeq,
      request = src.getRequest.fromProto(),
      data = if (src.getData != null) src.getData.toByteArray else Array.emptyByteArray)

    override def fromBytes[T: ClassTag](bytes: Array[Byte]): T = ???
  }
}

File: python/eggroll/core/command/command_model.py
#  Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

from urllib.parse import urlparse, parse_qs

from eggroll.core.base_model import RpcMessage
from eggroll.core.proto import command_pb2
from eggroll.core.utils import _map_and_listify, get_eggroll_bin_truncate_limit


class ErCommandRequest(RpcMessage):
    def __init__(self, id, uri: str, args: list = None, kwargs: dict = None):
        if args is None:
            args = []
        if kwargs is None:
            kwargs = {}
        self._id = id
        self._uri = uri
        self._args = args
        self._kwargs = kwargs

    def to_proto(self):
        return command_pb2.CommandRequest(id=self._id, uri=self._uri,
                                          args=self._args, kwargs=self._kwargs)

    @staticmethod
    def from_proto(pb_message):
        return ErCommandRequest(id=pb_message.id,
                                uri=pb_message.uri,
                                args=pb_message.args,
                                kwargs=pb_message.kwargs)

    @staticmethod
    def from_proto_string(pb_string):
        pb_message = command_pb2.CommandRequest()
        msg_len = pb_message.ParseFromString(pb_string)
        return ErCommandRequest.from_proto(pb_message)

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        return f'ErCommandRequest(id={self._id}, uri={self._uri}, ' \
               f'args=[{_map_and_listify(lambda v : v[:get_eggroll_bin_truncate_limit()], self._args)}, len={len(self._args)}], ' \
               f'kwargs=[***, len={len(self._kwargs)}])'


class ErCommandResponse(RpcMessage):
    def __init__(self, id, request: ErCommandRequest = None, results: list = None):
        if results is None:
            results = []
        self._id = id
        self._request = request
        self._results = results

    def to_proto(self):
        return command_pb2.CommandResponse(id=self._id,
                                           request=self._request.to_proto() if self._request else None,
                                           results=self._results)

    @staticmethod
    def from_proto(pb_message):
        return ErCommandResponse(id=pb_message.id,
                                 request=ErCommandRequest.from_proto(
                                         pb_message.request),
                                 results=pb_message.results)

    @staticmethod
    def from_proto_string(pb_string):
        pb_message = command_pb2.CommandResponse()
        msg_len = pb_message.ParseFromString(pb_string)
        return ErCommandResponse.from_proto(pb_message)

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        return f'ErCommandResponse(id={self._id}, request={repr(self._request)}, results=[{_map_and_listify(lambda v : v[:get_eggroll_bin_truncate_limit()], self._results)}, len={len(self._results)}]'


class ErService(object):
    def __init__(self, service_name, service_param_deserializers,
            service_result_serializers, route_to_class, route_to_method,
            call_based_instance, scope):
        self._service_name = service_name
        self._service_param_deserializers = service_param_deserializers
        self._service_result_serializers = service_result_serializers
        self._route_to_class = route_to_class
        self._route_to_method = route_to_method
        self._call_based_instance = call_based_instance
        self._scope = scope


class CommandURI(object):
    def __init__(self, uri_string='', command_request=None):
        if command_request:
            uri_string = getattr(command_request, "_uri")

        self._uri = uri_string
        self._prefix = ''
        self._parse_result = urlparse(self._uri)
        self._query_string = self._parse_result.query
        self._query_pairs = parse_qs(
                self._query_string)  # dict of (key: str, [value: str])

        if not self._query_pairs:
            self._query_pairs['route'] = uri_string

    def get_query_value(self, key: str):
        value = self._query_pairs[key]
        if value:
            return value

    def get_route(self):
        return self.get_query_value('route')

    def get_service_name(self):
        return "/".join([self._prefix, self._uri])

    def __repr__(self):
        return f'CommandURI(_uri={self._uri})'
